src/execution/window_executor.cpp

```diff
@@ -0,0 +1,1280 @@
+	auto rdata = FlatVector::GetData<double>(result);
+	for (idx_t i = 0; i < count; ++i, ++row_idx) {
+		int64_t denom = partition_end[i] - partition_begin[i];
+		double cume_dist = denom > 0 ? ((double)(peer_end[i] - partition_begin[i])) / denom : 0;
+		rdata[i] = cume_dist;
+	}
+}
+
+//===--------------------------------------------------------------------===//
+// WindowValueExecutor
+//===--------------------------------------------------------------------===//
+WindowValueExecutor::WindowValueExecutor(BoundWindowExpression &wexpr, ClientContext &context,
+                                         const idx_t payload_count, const ValidityMask &partition_mask,
+                                         const ValidityMask &order_mask)
+    : WindowExecutor(wexpr, context, payload_count, partition_mask, order_mask) {
+}
+
+WindowNtileExecutor::WindowNtileExecutor(BoundWindowExpression &wexpr, ClientContext &context,
+                                         const idx_t payload_count, const ValidityMask &partition_mask,
+                                         const ValidityMask &order_mask)
+    : WindowValueExecutor(wexpr, context, payload_count, partition_mask, order_mask) {
+}
+
+void WindowValueExecutor::Sink(DataChunk &input_chunk, const idx_t input_idx, const idx_t total_count) {
+	// Single pass over the input to produce the global data.
+	// Vectorisation for the win...
+
+	// Set up a validity mask for IGNORE NULLS
+	bool check_nulls = false;
+	if (wexpr.ignore_nulls) {
+		switch (wexpr.type) {
+		case ExpressionType::WINDOW_LEAD:
+		case ExpressionType::WINDOW_LAG:
+		case ExpressionType::WINDOW_FIRST_VALUE:
+		case ExpressionType::WINDOW_LAST_VALUE:
+		case ExpressionType::WINDOW_NTH_VALUE:
+			check_nulls = true;
+			break;
+		default:
+			break;
+		}
+	}
+
+	if (!wexpr.children.empty()) {
+		payload_chunk.Reset();
+		payload_executor.Execute(input_chunk, payload_chunk);
+		payload_chunk.Verify();
+		payload_collection.Append(payload_chunk, true);
+
+		// process payload chunks while they are still piping hot
+		if (check_nulls) {
+			const auto count = input_chunk.size();
+
+			UnifiedVectorFormat vdata;
+			payload_chunk.data[0].ToUnifiedFormat(count, vdata);
+			if (!vdata.validity.AllValid()) {
+				//	Lazily materialise the contents when we find the first NULL
+				if (ignore_nulls.AllValid()) {
+					ignore_nulls.Initialize(total_count);
+				}
+				// Write to the current position
+				if (input_idx % ValidityMask::BITS_PER_VALUE == 0) {
+					// If we are at the edge of an output entry, just copy the entries
+					auto dst = ignore_nulls.GetData() + ignore_nulls.EntryCount(input_idx);
+					auto src = vdata.validity.GetData();
+					for (auto entry_count = vdata.validity.EntryCount(count); entry_count-- > 0;) {
+						*dst++ = *src++;
+					}
+				} else {
+					// If not, we have ragged data and need to copy one bit at a time.
+					for (idx_t i = 0; i < count; ++i) {
```

Comment:
Could we write a test that covers this?

Reply:
This is covered by `test/sql/window/test_ignore_nulls.test` with vector size 2